### PR TITLE
Generate: fix flaky tests

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes two flaky tests:
1. `tests/models/clvp/test_modeling_clvp.py::ClvpDecoderTest::test_constrained_beam_search_generate` was failing at a ~3% rate because the range for which we were sampling tokens to force in constrained beam search was based on the random input_ids and sometimes invalid. 👉 fix = hard code the valid range based on `config.vocab_size`
2. `tests/models/marian/test_modeling_marian.py::MarianStandaloneDecoderModelTest::test_sample_generate` was failing at a ~0.1% rate because of a combination of poor logits processor setup + random input_ids 👉 fix = don't set up the logits processor with a combination that may lead to no valid continuation being possible

### CODE REVIEW

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1301,8 +1301,9 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         # set all nan values to 0.0
         scores[scores != scores] = 0.0
 
-        # set all inf values to max possible value
+        # set all +/-inf values to max/min possible value
         scores[scores == float("inf")] = torch.finfo(scores.dtype).max
+        scores[scores == float("-inf")] = torch.finfo(scores.dtype).min
```
Comment (gante) [inline src/transformers/generation/logits_process.py:1306]: This wasn't needed for the fix, but I noticed that it was missing :)

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -153,12 +157,16 @@ def _get_logits_processor_and_kwargs(
...
-                RepetitionPenaltyLogitsProcessor(process_kwargs["repetition_penalty"]),
-            ]
+            + [NoBadWordsLogitsProcessor(process_kwargs["bad_words_ids"], eos_token_id)]
+            + (
+                NoRepeatNGramLogitsProcessor(process_kwargs["no_repeat_ngram_size"])
+                if forced_bos_token_id is None and forced_eos_token_id is None
+                else []
+            )
+            + [RepetitionPenaltyLogitsProcessor(process_kwargs["repetition_penalty"])]
+            + [InfNanRemoveLogitsProcessor()]  # prevent flaky generation test failures
```
Comment (gante) [inline tests/generation/test_utils.py:167]: This one was being added appended before the `sample` call, and `generate` had `remove_invalid_values=True` (which triggers this logits processor).

Since `generate` appends logits processors passed with the `logits_processor` kwarg at the end of the list, this meant that prior to this change, `generate` was calling `InfNanRemoveLogitsProcessor` as the first logits processor, and `sample` as the last one. 

Order of operations matters in the logits processors! :)

**Review by amyeroberts — APPROVED** [review]
Thanks for the rapid fix! 🚀

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -123,9 +123,13 @@ def _get_logits_processor_and_kwargs(
         process_kwargs = {
             "min_length": input_length + 1 if max_length is None else max_length - 1,
             "bad_words_ids": [[1, 0]],
-            "no_repeat_ngram_size": 2,
             "repetition_penalty": 1.2,
+            "remove_invalid_values": True,
         }
+        # NoRepeatNGramLogitsProcessor + forced tokens may result in no valid continuations
```
Comment (gante) [inline tests/generation/test_utils.py:129]: This was the cause of the flakiness in the BART test (and very likely in other models as well!)

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -404,7 +407,6 @@ def _beam_search_generate(
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
             return_dict_in_generate=return_dict_in_generate,
-            remove_invalid_values=True,
```
Comment (gante) [inline tests/generation/test_utils.py:407]: This kwarg is now in `logits_process_kwargs`, two lines below

File: tests/generation/test_utils.py [hunk 5]
```diff
@@ -153,12 +157,16 @@ def _get_logits_processor_and_kwargs(
...
-                RepetitionPenaltyLogitsProcessor(process_kwargs["repetition_penalty"]),
-            ]
+            + [NoBadWordsLogitsProcessor(process_kwargs["bad_words_ids"], eos_token_id)]
+            + (
+                [NoRepeatNGramLogitsProcessor(process_kwargs["no_repeat_ngram_size"])]
+                if forced_bos_token_id is None and forced_eos_token_id is None
+                else []
+            )
+            + [RepetitionPenaltyLogitsProcessor(process_kwargs["repetition_penalty"])]
+            + [InfNanRemoveLogitsProcessor()]  # prevent flaky generation test failures
```
Comment (ydshieh) [inline tests/generation/test_utils.py:167]: probably extend the comment: by keeping the order of logit processors the same between `genearte` and `sample`

**Review by ydshieh — APPROVED** [review]
Thank you for taking care this!

(It seems to me there are 3 cases being fixed instead of 2: the extra one not mentioned is `fix the wrong order of logit processors`).
